### PR TITLE
feat(config): add `shuffle.files` and `shuffle.tests` options

### LINKED ISSUE #3568: `sequence.shuffle: true` does not respect the cache, runs test suites in random order
### Describe the bug

I want tests to run in random order, and I also want slower test suites to run first and quicker test suites to run last, but it seems impossible without implementing a custom `sequencer`.

With test suites/files being isolated, I don't see the point of also running the test suites/files in random order.

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-c5wvvm?file=test%2Fsuite1.test.ts,test%2Fslow.test.ts,test%2Fsuite2.test.ts,vite.config.ts,test%2Fsuite0.test.ts&initialPath=__vitest__/

1. run `npm run test:run` to build the cache
2. try `npm run test:run` multiple times, you'll see the test suites run in random order
3. set `shuffle` to `false`, run `npm run test:run`, notice that `test/slow.test.ts` starts first

### System Info

```shell
System:
    OS: Windows 10 10.0.22621
    CPU: (64) x64 AMD Ryzen Threadripper 3970X 32-Core Processor
    Memory: 17.54 GB / 31.86 GB
  Binaries:
    Node: 20.2.0 - C:\Program Files\nodejs\node.EXE
    npm: 9.6.7 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.22621.1778.0), Chromium (114.0.1823.43)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @vitejs/plugin-react: ^4.0.0 => 4.0.0
    @vitest/coverage-v8: ^0.32.0 => 0.32.0
    vite: ^4.3.5 => 4.3.9
    vitest: ^0.32.0 => 0.32.0
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by fenghan34**
### Description

- Add `shuffle.files` and `shuffle.tests` options, allowing users to shuffle tests while leveraging cache to run long-running tests earlier. 
- Closes #3568 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/node/config.ts [hunk 1]
```diff
@@ -441,6 +441,11 @@ export function resolveConfig(
     resolved.cache.dir = VitestCache.resolveCacheDir(resolved.root, resolved.cache.dir, resolved.name)
 
   resolved.sequence ??= {} as any
+  if (resolved.sequence.shuffle && typeof resolved.sequence.shuffle === 'object') {
+    const { files, tests } = resolved.sequence.shuffle
+    resolved.sequence.sequencer ??= files ? RandomSequencer : BaseSequencer
+    resolved.sequence.shuffle = files || tests
```
Comment (hi-ogawa) [inline packages/vitest/src/node/config.ts:447]: When `{ files: true, tests: false }`, wouldn't this shuffle test suites/cases?
I guess such pattern is quite unusual, but I thought `resolved.sequence.shuffle = tests` is simply fine here?

Comment (fenghan34) [reply packages/vitest/src/node/config.ts]: Good catch!! I've updated it.

**Review by hi-ogawa — APPROVED** [review]
Looks good!

File: test/core/test/cli-test.test.ts [hunk 2]
```diff
@@ -330,4 +330,13 @@ test('public parseCLI works correctly', () => {
       'color': true,
     },
   })
+
+  expect(parseCLI('vitest --sequence.shuffle.files --sequence.shuffle.tests=false')).toEqual({
```
Comment (sheremet-va) [inline test/core/test/cli-test.test.ts:334]: I would prefer if we had this in a separate test that tests all `shuffle` options. Please, don't use this specific test as it's only for the public `parseCLI` option, and it's already tested enough

Comment (fenghan34) [reply test/core/test/cli-test.test.ts]: I've applied your suggestion, thanks for letting me know that.

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**